src/libmongoc/tests/test-mongoc-client-side-encryption.c

```diff
@@ -680,7 +682,175 @@ test_datakey_and_double_encryption_creating_and_using (
+      ASSERT_OR_PRINT (client_encryption, error);
+
+      mongoc_client_encryption_opts_destroy (client_encryption_opts);
+      bson_destroy (kms_providers);
+      bson_destroy (tls_opts);
+   }
+
+   /* Using client_encryption, create a data key with a local KMS provider and
+    * the following custom key material: */
+   {
+      const char key_material[] =
+         "xPTAjBRG5JiPm+d3fj6XLi2q5DMXUS/"
+         "f1f+SMAlhhwkhDRL0kr8r9GDLIGTAGlvC+HVjSIgdL+"
+         "RKwZCvpXSyxTICWSXTUYsWYPyu3IoHbuBZdmw2faM3WhcRIgbMReU5";
+      uint8_t data[96];
+      mongoc_client_encryption_datakey_opts_t *datakey_opts =
+         mongoc_client_encryption_datakey_opts_new ();
+      bson_value_t keyid;
+
+      BSON_ASSERT (
+         mcommon_b64_pton (key_material, data, sizeof (key_material)) == 96);
+
+      mongoc_client_encryption_datakey_opts_set_keymaterial (
+         datakey_opts, data, sizeof (data));
+
+      ASSERT_OR_PRINT (
+         mongoc_client_encryption_create_key (
+            client_encryption, "local", datakey_opts, &keyid, &error),
+         error);
+
+      ASSERT (keyid.value_type == BSON_TYPE_BINARY);
+      ASSERT (keyid.value.v_binary.subtype == BSON_SUBTYPE_UUID);
+      ASSERT (keyid.value.v_binary.data_len != 0);
+
+      mongoc_client_encryption_datakey_opts_destroy (datakey_opts);
+      bson_value_destroy (&keyid);
+   }
+
+   /* Find the resulting key document in keyvault.datakeys, save a copy of the
+    * key document, then remove the key document from the collection. */
+   {
+      mongoc_collection_t *const datakeys =
+         mongoc_client_get_collection (client, "keyvault", "datakeys");
+      mongoc_cursor_t *cursor = mongoc_collection_find_with_opts (
+         datakeys, tmp_bson ("{}"), NULL /* opts */, NULL /* read prefs */);
+      const bson_t *bson;
+
+      ASSERT (mongoc_cursor_next (cursor, &bson));
+      bson_copy_to (bson, &datakey);
+      mongoc_cursor_destroy (cursor);
+
+      (void) mongoc_collection_drop (datakeys, &error);
+      mongoc_collection_destroy (datakeys);
+   }
+
+   /* Replace the _id field in the copied key document with a UUID with base64
+    * value AAAAAAAAAAAAAAAAAAAAAA== (16 bytes all equal to 0x00) and insert the
+    * modified key document into keyvault.datakeys. */
+   {
+      mongoc_collection_t *const datakeys =
+         mongoc_client_get_collection (client, "keyvault", "datakeys");
+      bson_t modified_datakey = BSON_INITIALIZER;
+      uint8_t bytes[16] = {0};
+
+      bson_copy_to_excluding_noinit (&datakey, &modified_datakey, "_id", NULL);
+      BSON_ASSERT (BSON_APPEND_BINARY (
+         &modified_datakey, "_id", BSON_SUBTYPE_UUID, bytes, sizeof (bytes)));
+
+      ASSERT_OR_PRINT (mongoc_collection_insert_one (
+                          datakeys, &modified_datakey, NULL, NULL, &error),
+                       error);
+
+      mongoc_collection_destroy (datakeys);
+      bson_destroy (&modified_datakey);
```

Comment:
Use majority write concern to insert. Majority read concern is used to find keys. If majority write concern is not used, the find may not see the inserted document if it has not replicated to a majority of nodes in a replica set. Also suggested on the specification https://github.com/mongodb/specifications/pull/1213/files#r875179801

Here is a sample of setting majority write concern:
```c
      mongoc_write_concern_t *wc;
      bson_t insert_opts = BSON_INITIALIZER;

      wc = mongoc_write_concern_new ();
      mongoc_write_concern_set_w (wc, MONGOC_WRITE_CONCERN_W_MAJORITY);
      mongoc_write_concern_append (wc, &insert_opts);

      ASSERT_OR_PRINT (
         mongoc_collection_insert_one (
            datakeys, &modified_datakey, NULL, &insert_opts, &error),
         error);

      mongoc_write_concern_destroy (wc);
      bson_destroy (&insert_opts);
```